File: spec/java/src/io/kaitai/struct/specwrite/TestExpr2.java
package io.kaitai.struct.specwrite;

import io.kaitai.struct.ByteBufferKaitaiStream;
import io.kaitai.struct.KaitaiStruct.ReadWrite;
import io.kaitai.struct.testwrite.Expr2;
import org.testng.annotations.Test;
import java.nio.charset.StandardCharsets;
import static org.testng.Assert.*;

public class TestExpr2 extends CommonSpec {
    @Override
    protected Class<? extends ReadWrite> getStructClass() {
        return Expr2.class;
    }

    @Override
    protected String getSrcFilename() {
        return "str_encodings.bin";
    }

    @Test
    public void testEdit() throws Exception {
        Expr2 r = Expr2.fromFile(SRC_DIR + getSrcFilename());
        r._read();

        int oldLenMod = r.str2().lenMod();
        int oldStr2RestAvg = r.str2().rest().avg();

        r.str2().setStr("Kaitai Struct カ");
        r.str2()._invalidateLenMod();

        int str2Size = r.str2().str().getBytes(StandardCharsets.UTF_8).length;
        r.str2().setLenOrig(str2Size + 3);
        assertNotEquals(r.str2().lenMod(), oldLenMod);
        assertIntEquals(r.str2().lenMod(), str2Size);

        r.str2().rest().setByte0(0xfa);
        r.str2().rest()._invalidateAvg();
        r.str2().rest().setByte1(0x44);
        r.str2().rest().setByte2(0x88);
        assertNotEquals(r.str2().rest().avg(), oldStr2RestAvg);

        // see <CommonSpec>.testReadWriteRoundtrip
        Object origDump = dumpStruct(r);

        ByteBufferKaitaiStream newIo = new ByteBufferKaitaiStream(
            2 + // str1.len_orig
            r.str1().lenMod() + // str1.str
            3 + // str1.rest
            2 + // str2.len_orig
            r.str2().lenMod() + // str2.str
            3 // str2.rest
        );
        r._write(newIo);
        newIo.seek(0);

        Expr2 newR = new Expr2(newIo);
        newR._read();

        Object newDump = dumpStruct(newR);

        assertEquals(newDump, origDump);
    }
}

File: spec/python/specwrite/test_expr_2.py
# -*- coding: utf-8 -*-
import io
import unittest
from kaitaistruct import KaitaiStream
from specwrite.common_spec import CommonSpec

from testwrite.expr_2 import Expr2

class TestExpr2(CommonSpec.Base):
    def __init__(self, *args, **kwargs):
        super(TestExpr2, self).__init__(*args, **kwargs)
        self.struct_class = Expr2
        self.src_filename = 'src/str_encodings.bin'

    def test_edit(self):
        r = Expr2.from_file(self.src_filename)
        r._read()

        old_len_mod = r.str2.len_mod
        old_str2_rest_avg = r.str2.rest.avg

        r.str2.str = u"Kaitai Struct カ"
        r.str2._invalidate_len_mod()

        str2_size = len(r.str2.str.encode(u"UTF-8"))
        r.str2.len_orig = str2_size + 3
        self.assertNotEqual(r.str2.len_mod, old_len_mod)
        self.assertEqual(r.str2.len_mod, str2_size)

        r.str2.rest.byte0 = 0xfa
        r.str2.rest._invalidate_avg()
        r.str2.rest.byte1 = 0x44
        r.str2.rest.byte2 = 0x88
        self.assertNotEqual(r.str2.rest.avg, old_str2_rest_avg)

        # see <CommonSpec.Base>.test_read_write_roundtrip
        orig_dump = CommonSpec.Base.dump_struct(r)

        new_io = KaitaiStream(io.BytesIO(bytearray(
            2 +  # str1.len_orig
            r.str1.len_mod +  # str1.str
            3 +  # str1.rest
            2 +  # str2.len_orig
            r.str2.len_mod +  # str2.str
            3  # str2.rest
        )))
        r._write(new_io)
        new_io.seek(0)

        new_r = Expr2(new_io)
        new_r._read()

        new_dump = CommonSpec.Base.dump_struct(new_r)

        self.assertEqual(new_dump, orig_dump)
